### PR TITLE
Mask loss at document boundaries

### PR DESCRIPTION
**Pull request by francesco-bertolotti**
This is probably not a significant issue, but it seems like the more correct way to handle document packing.

Currently, when training with document packing, the last token of one document is used to predict the first token of the next document.

Because the tokenizer always appends an EOS token and prepends a BOS token, the model is effectively trained to predict a BOS token after every EOS token. Given this behavior, the current implementation does not introduce an issue. However, without these special tokens, the last token of one document would be trained to predict the first token of an unrelated document, which is clearly undesirable.

Although this is not a problem in the current setup, I believe the correct approach is to ignore the loss for the last token of each document. This avoids learning cross-document transitions and aligns the training objective with document boundaries. Additionally, implementing this requires only a one-line change.

---

This PR probably breaks some tests. Before fixing those too, I would like to know what do you think about this.

### CODE REVIEW

File: torchtitan/hf_datasets/text_datasets.py [hunk 1]
```diff
@@ -160,7 +160,10 @@ def __iter__(self):
                     ]
 
                     input = x[:-1]
-                    label = x[1:]
+                    label = x[1:].clone()
+                    # Mask EOS tokens in the label to avoid predicting from
+                    # next-document context.
+                    label[input == self._tokenizer.eos_id] = IGNORE_INDEX
```
Comment (tianyu-l) [inline torchtitan/hf_datasets/text_datasets.py]: Nice catch!

Instead of doing this, where we always "waste" the training when input is `eos_id` as we do `input`, `label` split on the packed buffer, how about let's follow the `ChatDataloader` and do `input`, `label` split per each sample
https://github.com/pytorch/torchtitan/blob/main/torchtitan/hf_datasets/text_datasets.py#L482-L483

**Review by tianyu-l — CHANGES_REQUESTED** [review]

**Comment by francesco-bertolotti** [comment]
Hi @tianyu-l 

> Instead of doing this, where we always "waste" the training when input is `eos_id` as we do input/label split on the packed buffer, how about let's follow the ChatDataloader and do input/label split per each sample?

Seems reasonable to me. With this change, the shift now happens at the document level rather than the row level. I’ve also added a few tests with some basic checks.

A few things to be aware of:

* `EOS` does not appear in the input IDs.
* `BOS` does not appear in the label IDs.
* Resuming a run from this commit will result in a short period of inconsistent behavior, since the saved buffers in the state dict will have both `BOS` and `EOS` in the input/label IDs. I think this makes the change somewhat `breaking`, although it is unlikely to cause any real issues.
* Unless weight tying is applied, the `EOS` input embedding is not trained, since `EOS` never appears in the input IDs.

**Review by tianyu-l — APPROVED** [review]
lgtm -- the numerics drift in CI is expected, could you update the loss files?

@felipemello1 to take another look

File: torchtitan/hf_datasets/text_datasets.py [hunk 2]
```diff
@@ -186,13 +187,23 @@ def reloop(self) -> None:
...
+            # raw, unshifted token stream in 'inputs_buffer'. It cannot be split
+            # into input/label pairs here without reintroducing the cross-document
+            # targets this shift exists to avoid, so drop the partial sample and
+            # resume at the next document boundary.
             logger.warning(
-                "Checkpoint missing 'positions_buffer'. Falling back to empty buffer. "
-                "RoPE positions may be incorrect with block_causal attention."
+                "Checkpoint missing 'labels_buffer'. Dropping the buffered partial "
+                "sample; iteration resumes at the next document boundary."
             )
```
Comment (felipemello1) [inline torchtitan/hf_datasets/text_datasets.py]: i dont think that we need backwards compatibility. All datasets from here on should have labels buffer and positions buffer, right? should we just delete it @francesco-bertolotti ?

@tianyu-l  can you confirm.

**Review by felipemello1 — APPROVED** [review]
lgtm! thank you!

File: torchtitan/hf_datasets/text_datasets.py (hunk 2 above)
Comment (tianyu-l) [reply torchtitan/hf_datasets/text_datasets.py]: yeah BC is secondary concern, I don't care that much

Comment (francesco-bertolotti) [reply torchtitan/hf_datasets/text_datasets.py]: Sounds reasonable

**Comment by francesco-bertolotti** [comment]
I’ve updated `text_datasets.py` to address the backward compatibility issue.

Regarding the numerics drift in CI, I have updated `tests/assets/losses/llama3_cuda.txt ` according to what was written in https://github.com/pytorch/torchtitan/actions/runs/31304921466/job/93296866246?pr=4075

**Comment by tianyu-l** [comment]
seems there are still CI errors

**Comment by francesco-bertolotti** [comment]
Sorry, Let's hope this time works.

- Updated `tests/assets/losses/qwen3_moe_cuda.txt` according to shown numerics in CI
- Added empty label buffer to state dict in tests/unit_tests/test_dataset_checkpointing.py
